pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,391 @@
+// 10.  Old paths are removed from the user-visible portion of the target directory
+// 11.  The sentinel file modification and access times are updated (file is created if it does not
+//      already exist)
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	cleanPayload, err := validatePayload(payload)
+	if err != nil {
+		glog.Errorf("%s: invalid payload: %v", w.LogContext, err)
+		return err
+	}
+
+	pathsToRemove, err := w.pathsToRemove(cleanPayload)
+	if err != nil {
+		glog.Errorf("%s: error determining files to prune: %v", w.LogContext, err)
+		return err
+	}
+
+	if should, err := w.shouldWritePayload(cleanPayload); err != nil {
+		glog.Errorf("%s: error determining whether payload should be written to disk: %v", w.LogContext, err)
+		return err
+	} else if !should && len(pathsToRemove) == 0 {
+		glog.V(5).Infof("%s: no update required for target directory %v", w.LogContext, w.TargetDir)
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		glog.V(5).Infof("%s: error creating new ts data directory: %v", w.LogContext, err)
+		return err
+	}
+
+	if err = w.writePayloadToDir(cleanPayload, tsDir); err != nil {
+		glog.Errorf("%s: error writing payload to ts data directory %s: %v", w.LogContext, tsDir, err)
+		return err
+	}
+
+	if err = w.createUserVisibleFiles(cleanPayload); err != nil {
+		glog.Errorf("%s: error creating visible symlinks in %s: %v", w.LogContext, w.TargetDir, err)
+		return err
+	}
+
+	_, tsDirName := filepath.Split(tsDir)
+	dataDirPath := path.Join(w.TargetDir, dataDirName)
+	oldTsDir, err := os.Readlink(dataDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("%s: error reading link for data directory: %v", w.LogContext, err)
+		return err
+	}
+
+	newDataDirPath := path.Join(w.TargetDir, newDataDirName)
+	if err = os.Symlink(tsDirName, newDataDirPath); err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error creating symbolic link for atomic update: %v", w.LogContext, err)
+		return err
+	}
+
+	if err = os.Rename(newDataDirPath, dataDirPath); err != nil {
+		os.Remove(newDataDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error renaming symbolic link for data directory %s: %v", w.LogContext, newDataDirPath, err)
+		return err
+	}
+
+	if len(oldTsDir) > 0 {
+		if err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir)); err != nil {
+			glog.Errorf("%s: error removing old data directory %s: %v", w.LogContext, oldTsDir, err)
+			return err
+		}
+	}
+
+	if err = w.removeUserVisiblePaths(pathsToRemove); err != nil {
+		glog.Errorf("%s: error pruning old visible symlinks: %v", w.LogContext, err)
+		return err
+	}
+
+	return w.touchSentinelFile()
+}
+
+// validatePayload returns an error if any path in the payload  returns a copy of the payload with the paths cleaned.
+func validatePayload(payload map[string][]byte) (map[string][]byte, error) {
+	cleanPayload := make(map[string][]byte)
+	for k, content := range payload {
+		if strings.Contains(k, "..") {
```

Comment:
I see - you want to prevent ../ - I think you should be explicit.  `^..` is not allowed nor is any segment being `..`